be/src/olap/short_key_index.cpp

```diff
@@ -0,0 +1,153 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "olap/short_key_index.h"
+
+#include <string>
+
+#include "util/coding.h"
+#include "gutil/strings/substitute.h"
+
+using strings::Substitute;
+
+namespace doris {
+
+Status ShortKeyIndexBuilder::init() {
+    _key_length = 0;
+    for (auto& key : _key_fields) {
+        // +1 for null bit, we should improve this some day
+        _key_length += key.index_length + 1;
+    }
+    _tmp_key_buf.resize(_key_length);
+    return Status::OK();
+}
+
+Status ShortKeyIndexBuilder::add_item(const RowCursor& short_key, uint32_t block_id) {
+    char* key_buf = (char*)_tmp_key_buf.data();
+
+    size_t offset = 0;
+    for (size_t i = 0; i < _key_fields.size(); i++) {
+        short_key.write_index_by_index(i, key_buf + offset);
+        offset += short_key.get_index_size(i) + 1;
+    }
+
+    _index_buf.append(key_buf, _key_length);
+    put_fixed32_le(&_index_buf, block_id);
```

Comment:
I think can add a config or a member variable to determine whether to write the block id. And This information can be parsed from header() and be self-explained.
because block id will be discarded in V2. Then the code will not be modified again.

Reply:
Because we reuse MemIndex class, which should be compatible with old version format. If we do this, we should write a new class MemIndexV2.

So, Let's do this later. I think we can do this after we pass the smoke test